### ISSUE TITLE
Re-enable MyPy

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#19891]
# Why Mypy re-enable

For a few weeks MyPy checks have been disabled after the switch to Python 3.7 (per https://github.com/apache/airflow/pull/19317).

We should, however, re-enable it back as it is very useful in catching a number of mistakes.

# How does it work

We 've re-added the mypy pre-commit now - with mypy bumped to 0.910. This version detects far more errors and we should fix them all before we switch the CI check back.

* mypy will be running for incremental changes in pre-commit, same as before. This will enable incremental fixes of the code changed by committers who use pre-commits locally

* mypy on CI runs in non-failing mode. When the main pre-commit check is run, mypy is disabled, but then it is run as a separate step (which does not fail but will show the result of running mypy on all our code). This will enable us to track the progress of fixes

# Can I help with the effort, you ask? 

We started concerted effort now and incrementally fix all the mypy incompatibilities - ideally package/by/package to avoid huge code reviews. We'd really appreciate a number of people to contribute, so that we can re-enable mypy back fully and quickly :).

# How can I help?

What you need is:

* checkout `main`
* `./breeeze build-image`
* `pip install pre-commit`
* `pre-commit install`

This will enable automated checks for when you do a regular contribution.  When you make your change, any MyPy issues will be reporteed and you need to fix them all to commit. You can also commit with `--no-verify` flag to skip that, bu, well, if you can improve airlfow a little - why not?  

# How can I help more ? 

You can add PRs that are fixing whole packages, without contributing features or bugfixes. Please refer to this issue #19891 and ideally comment below in the issue that you want to take care of a package (to avoid duplicate work).

An easy way to run MyPy check for package can be done either from the host:
```
find DIRECTORY -name "*.py" | xargs pre-commit run mypy --files
```

or from ./breeze shell:
```
mypy --namespace-packages DIRECTORY
```

# Current list of mypy PRs:

https://github.com/apache/airflow/pulls?q=is%3Aopen+is%3Apr+label%3Amypy

# Remaining packages

Here is the list of remaining packages to be "mypy compliant" generated with:

```
pre-commit run mypy --all-files 2>&1 | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | grep "error:" | sort | awk 'FS=":" { print $1 }' | xargs dirname | sort | uniq -c | xargs -n 2 printf "* [ ] (%4d) %s\n"
```

* [ ] (   1) airflow/api/common/experimental
* [ ] (   1) airflow/contrib/sensors
* [ ] (   1) airflow/example_dags
* [ ] (   1) airflow/jobs
* [ ] (   4) airflow/models
* [ ] (   1) airflow/providers/microsoft/winrm/hooks
* [ ] (   1) airflow/providers/ssh/hooks
* [ ] (   1) tests/providers/amazon/aws/hooks
* [ ] (   1) tests/providers/google/cloud/hooks

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by josh-fell** [comment]
Love this! Happy to help here. Do you suspect a large number of fixes needed, and if so, should we coordinate over a Slack channel or just keep it to this issue?

**Comment by potiuk** [comment]
Yeah we have about 800 issues :( . I think what I might do is to generate a list of packages and we can subscribe to them here in comments and check them when they are done.

**Comment by potiuk** [comment]
There you are @josh-fell - we have 992 unique issues

**Comment by josh-fell** [comment]
Oof. Let the fun begin!

Naturally, I'll sign up for all of the `*/example_dags`. And I'll take the Azure ones too if you don't mind.

**Comment by potiuk** [comment]
I will start with "airflow" part :)

**Comment by potiuk** [comment]
But we need to merge https://github.com/apache/airflow/pull/19890 first

**Comment by ashb** [comment]
Ouch. I guess some/many of those are from upgrading Mypy and it adding new better checks, rather than us having broken things, right?

**Comment by potiuk** [comment]
> Ouch. I guess some/many of those are from upgrading Mypy and it adding new better checks, rather than us having broken things, right?

I am not sure. Just from the number of thos my guess is we simply missed a lot of errors via workarounding lack of namespace support (we had to convert files to packages and pass packages to mypy not files and I think it had bad side-effects).

Also I am not sure yet if mypy will detect all errors in " consistent" way with teh "parallel support from pre-commit" - it might be that when you pass all files in one run, it will detect more errors.  I will check it soon to see - but it is not needed to start fixing.

We might want to switch to this non-parallel mode, though it will be even more pain for local development (but then we might finally switch to mypyd support).

**Comment by potiuk** [comment]
Just checked. 

Parallel pre-commit is useless for `mypy` - seems that it will not speedup the tests anyway as mypy pretty much loads as many python code as it reaches out from the files, and with our code structure, any import to airflow will pretty much load everything always. So when we run MyPy in parallel, My 16 cores are fully busy and produce the same result in 2m6s as single busy core  in 2m13s. Results are the same. 

I will turn on serial mode :).

Serial:
```
[jarek:~/code/airflow] enable-mypy-as-non-failing-step(+3/-4)+ 8s ± pre-commit run mypy --all-files 2>&1 | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | grep "error:" | sort | awk 'FS=" " { print $1 }' | uniq > /tmp/res-serial.out 

[jarek:~/code/airflow] enable-mypy-as-non-failing-step(+3/-4)+ 2m13s ± wc /tmp/res-serial.out 
  963   963 48752 /tmp/res-serial.out
```

Paralel:
```
[jarek:~/code/airflow] enable-mypy-as-non-failing-step(+3/-4)+ ± pre-commit run mypy --all-files 2>&1 | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | grep "error:" | sort | awk 'FS=" " { print $1 }' | uniq > /tmp/res-paralle.out

[jarek:~/code/airflow] enable-mypy-as-non-failing-step(+1/-2)+ 2m6s ± wc /tmp/res-paralle.out 
  963   963 48752 /tmp/res-paralle.out
[jarek:~/code/airflow] enable-mypy-as-non-failing-step(+1/-2)+ ± 
```

**Comment by potiuk** [comment]
Opened since 2015: https://github.com/python/mypy/issues/933

**Comment by potiuk** [comment]
BTW. @josh-fell  and others -  I already have quite a number of fixes - semi-automated - (but not a lot of them are verified  in https://github.com/apache/airflow/pull/19334 - so let me take a look and maybe some packages/parts will be cleanly applicable and we can get rid of many issues - but I want to merge this one first and the apply what I have there package by package (or maybe group of packages0

**Comment by uranusjr** [comment]
I just read a few PRs on this, and the general impression I’m getting is we are using `type: ignore` way too much. I feel we should try a bit harder to fix the actual typing errors instead of liberally slapping on ignores everywhere.

**Comment by potiuk** [comment]
> I just read a few PRs on this, and the general impression I’m getting is we are using `type: ignore` way too much. I feel we should try a bit harder to fix the actual typing errors instead of liberally slapping on ignores everywhere.

Agree.

**Comment by potiuk** [comment]
Those PRs were just  "Quick" extract of what I've done before really quickly - but I agreee we should rather refactore some of the "core" code because it is often done "badly" using shortcuts and very loose approach to types (reusing same variables for different types freely for example).

I hope more peopel will join and it will take just a short while to complete .

**Comment by potiuk** [comment]
cc: @khalidmammadov

**Comment by khalidmammadov** [comment]
I can take a look to tests/* folder if it's ok?

**Comment by potiuk** [comment]
Perfect!

**Comment by Bowrna** [comment]
I will check this issue, understand it and start making contributions.

**Comment by khalidmammadov** [comment]
As a reference/check, this is how I get the list of issues that needs fixing in test/* folder. 
I first created requirement file to install required packages (inline with latest changes into repo)
```
[venv_airflow] ~/dev/oss/airflow (main*)$ cat ../mypy_requirements.txt 
mypy==0.910
types-croniter
types-docutils
types-freezegun
types-paramiko
types-protobuf
types-python-dateutil
types-python-slugify
types-pytz
types-redis
types-requests
types-setuptools
types-termcolor
types-tabulate
types-Markdown
types-PyMySQL
types-PyYAML

pip install -r ../mypy_requirements.txt

# and run below
[venv_airflow] ~/dev/oss/airflow (main*)$ mypy tests/
airflow/utils/weekday.py:62: error: Argument 1 to <set> has incompatible type "WeekDay"; expected "str"
                    week_day = {week_day}
                                ^
airflow/utils/edgemodifier.py:41: error: Incompatible default for argument "label" (default has type "None", argument has type "str")
        def __init__(self, label: str = None):
..........       
                   ^
tests/plugins/test_plugins_manager.py:45: error: Incompatible types in assignment (expression has type "str", variable has type Module)
            importlib_metadata = 'importlib.metadata'
                                 ^
Found 426 errors in 96 files (checked 1206 source files)
```

@potiuk does it sound reasonable or shall I use pre-commit run mypy ...  (which I did try but didnt get list and didnt look into it too much TBH)

**Comment by potiuk** [comment]
> @potiuk does it sound reasonable or shall I use pre-commit run mypy ... (which I did try but didnt get list and didnt look into it too much TBH)

I think it's better to do it inside `breeze` because there you have ALL dependencies installed in the latest "constraints-main" versions - so you are 100% sure that you have the same dependencies, and the problem is that mypy will show different errors depending whcih dependencies you have installed (because for example some dependency might have added type hinting in a different version). This is why at least `./breeze` is recommended as "common" execution environment for tests (that's the main purpose of Breeze anyway).

You do not have to run pre-commits. But at least emulating what they do is  useful.

Precommits under the hood do this  (at least for the `airflow` test - the namespace packages are not used for chart and docs - because they got brokent if they do) 

```
mypy --namespace-packages file file file ...
```

I am not 100% sure if just providing a package (as you did) will yield the same results. I think in this case `mypy` will only check "published" objectsi in `__init__.py` or something similar (but here I am not sure)

We used to do it  (we converted files to packages) to overcome mypy's poor support of namespace packages in 0.770, and by seing that after ~3 weeks we have 992 errors that I see now with mypy 0.990 I am afraid part of the problem was that we used packages rather than files. But I am not at all sure of it - those might be simply because mypy 0.990 has gotten so much better since (this is quite possible as 0.770 is > 1 year old)/

 You can check if it is the case now and let me know if you have different results if you try `package' vs. `module` approach) 

So at least check if when you do `mypy --namespace-packages test` and `find test | xargs mypy --namespace-packaes`  in ./breeze, you get the same results. My "hunch" is that they will differ.  But if you enter breeze and run `find test | xargs mypy --namespace-packaes` - this is what `pre-commit` actually does.

**Comment by khalidmammadov** [comment]
@potiuk I run below inside ./breeze. Looks like does not make much difference. 
mypy --namespace-packages tests/
# Reported:
# Found 145 errors in 32 files (checked 1208 source files)

find tests/ -name "*.py" | xargs mypy --namespace-packages
# Reported
# Found 145 errors in 32 files (checked 1208 source files)

**Comment by potiuk** [comment]
Sounds GOOD. So @ashb  answering your earlier question from https://github.com/apache/airflow/issues/19891#issuecomment-982921547 - it looks like all the 900+ errors ARE coming from migrating to the new mypy.

We should have done that a loooong time ago :).

**Comment by potiuk** [comment]
I updated the description with the "in-breeze" command and more ogranised approach in general (thanks @khalidmammadov)!

**Comment by josh-fell** [comment]
I took a crack at `airflow/decorators` but 1 error is eluding me (see #20034).

Anyone grab `airflow/providers/salesforce` yet?

**Comment by potiuk** [comment]
> Anyone grab `airflow/providers/salesforce` yet?

Feel free! . few more merges and I am going to refresh the list to see where we are, on top of manual updatees.

**Comment by josh-fell** [comment]
@potiuk Perhaps I should hold off based on [this comment](https://github.com/apache/airflow/pull/20034#discussion_r762427587)? Or maybe just start tackling `airflow/providers/**/hooks` in the meantime? WDYT?

That comment just raises concern about not fixing `BaseOperator` first and having some trickle effect we'll have to fix again in operators.

**Comment by potiuk** [comment]
Likely. Possibly reworking tests is better idea . Some overlap is bound to happen, but I am not sure if there are just a few things that we will have to redo, it's a big problem. We have  900 errors to fix  :)

**Comment by josh-fell** [comment]
@khalidmammadov You seem to be cruising on the `tests/*` set but let me know if there are any you'd like some help on or take off your plate until we get a green light on cranking through the providers.

**Comment by potiuk** [comment]
@josh-fell  - start from the end :D

**Comment by khalidmammadov** [comment]
@josh-fell Sure things, I have started from the top of tests and now on and doing tests/providers. As @potiuk mentioned, you can start from other end (www) and so we can get it done soon. Thanks

**Comment by khalidmammadov** [comment]
I think we done with tests/* folder. 
Starting to look into airflow/www/security if not started by anyone.

@potiuk BTW, my latest run reports:
**Found 808 errors in 212 files (checked 1820 source files)**

**Comment by potiuk** [comment]
cool. I will refresh the list from scratch tomorrow :).

**Comment by josh-fell** [comment]
I started looking into `dev`. Let me know if someone else has already though.

**Comment by potiuk** [comment]
All right I regenerated the current list of packages, and it went a bit down :).

Not every package we "cleaned up" is gone from the list  because as we clean-up and add Types in the "common" parts - new errors might appear in the packages that are using them.

From what i checked in "Breeze" we have still 952 errors  - but I think it should not discourage us. In the initial period I believe it will fluctuate up and down because of the above phenomena, but I hope very soon the numbers will start going down quickly!

**Comment by potiuk** [comment]
And the faster we fix the common packages, the faster the "dependent" packages will stabilize and the numbers will go down :)

**Comment by ashb** [comment]
> And the faster we fix the common packages, the faster the "dependent" packages will stabilize and the numbers will go down :)

I've also noticed that MyPy gives different results based on if you just check a module in isolation or the whole code base (i.e. `mypy airflow/models/dag.py` could say 0 errors, but `mypy airflow` could still list errors in dag.py. Fun)

**Comment by subkanthi** [comment]
Fixes mypy in airflow/providers/neo4j and airflow/providers/influxdb

https://github.com/apache/airflow/pull/20189/files

**Comment by subkanthi** [comment]
I come across this error a few times, is there a fix for this
Module "functools" has no attribute "cached_property"

**Comment by subkanthi** [comment]
Fix mypy in

- airflow/providers/apache/spark/hooks

https://github.com/apache/airflow/pull/20190

**Comment by josh-fell** [comment]
> I come across this error a few times, is there a fix for this `Module "functools" has no attribute "cached_property"`

You can use the `# type: ignore[attr-defined]` marker so Mypy ignores that specific error.

**Comment by uranusjr** [comment]
A better fix is

```python
if sys.version_info >= (3, 8):
    from functools import cached_property
else:
    from cached_property import cached_property
```

This is the solution used in `airflow.compat.functools`, which should be used by “proper” Airflow code. But providers unfortunately cannot use it (yet).

**Comment by josh-fell** [comment]
> A better fix is
> 
> ```python
> if sys.version_info >= (3, 8):
>     from functools import cached_property
> else:
>     from cached_property import cached_property
> ```
> 
> This is the solution used in `airflow.compat.functools`, which should be used by “proper” Airflow code. But providers unfortunately cannot use it (yet). Since `cached_property` is a single-use decorator, simply ignoring the error is also considered acceptable.

@uranusjr My recent mypy fix for Azure used that marker in 'airflow/providers/microsoft/azure/log/wasb_task_handler.py'. Happy to implement your suggestion instead if you feel it's best.

**Comment by ashb** [comment]
> > A better fix is
> > ```python
> > if sys.version_info >= (3, 8):
> >     from functools import cached_property
> > else:
> >     from cached_property import cached_property
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is the solution used in `airflow.compat.functools`, which should be used by “proper” Airflow code. But providers unfortunately cannot use it (yet). Since `cached_property` is a single-use decorator, simply ignoring the error is also considered acceptable.
> 
> @uranusjr My recent mypy fix for Azure used that marker in `airflow/providers/microsoft/azure/log/wasb_task_handler.py`. Happy to implement your suggestion instead if you feel it's best.

Yes please @josh-fell , we can't use `airflow.compat.functools` in providers, that module only got added in 2.2.0, so if we use that in providers it would need a min airflow version of 2.2.0.

**Comment by josh-fell** [comment]
> > > A better fix is
> > > ```python
> > > if sys.version_info >= (3, 8):
> > >     from functools import cached_property
> > > else:
> > >     from cached_property import cached_property
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This is the solution used in `airflow.compat.functools`, which should be used by “proper” Airflow code. But providers unfortunately cannot use it (yet). Since `cached_property` is a single-use decorator, simply ignoring the error is also considered acceptable.
> > 
> > 
> > @uranusjr My recent mypy fix for Azure used that marker in `airflow/providers/microsoft/azure/log/wasb_task_handler.py`. Happy to implement your suggestion instead if you feel it's best.
> 
> Yes please @josh-fell , we can't use `airflow.compat.functools` in providers, that module only got added in 2.2.0, so if we use that in providers it would need a min airflow version of 2.2.0.

On it. Thanks @ashb. I'll add the version check instead of using the `ignore[attr-defined]` comment that's there now.

**Comment by potiuk** [comment]
> Yes please @josh-fell , we can't use `airflow.compat.functools` in providers, that module only got added in 2.2.0, so if we use that in providers it would need a min airflow version of 2.2.0.

And @josh-fell - in case you asked - if you try to use the functools, the CI would have failed in the "install providers on 2.1" CI test :).

**Comment by subkanthi** [comment]
Just FYI - providers/google/cloud/hooks seems to be one of the biggest ones - 192 errors

**Comment by kazanzhy** [comment]
Fix mypy in

* airflow/providers/snowflake

#20212

**Comment by potiuk** [comment]
@subkanthi  - I will take the google one :scream:

**Comment by potiuk** [comment]
Just one corretion - it looks like 294 errors when I run it :)

**Comment by potiuk** [comment]
OK - I have quite a few google-provider related fixes in progress - I need to get them merged  first because I fixed some "related" errors in few of them and I am startting to see the errors that I fixed in previous PRs. Looking forward to reviews

**Comment by potiuk** [comment]
I added `mypy` label (temporarily) to keep track of all the opened MyPy PRs https://github.com/apache/airflow/pulls?q=is%3Aopen+is%3Apr+label%3Amypy

**Comment by josh-fell** [comment]
I can pickup `docs/ext/docs_build` and `scripts/in_container` next.

**Comment by josh-fell** [comment]
FYI @potiuk #20261 takes care of both `dev/` and `dev/provider_packages/` in the remaining packages list.

**Comment by potiuk** [comment]
Yeah. Good point @josh-fell .  Marked. 

But also I will keep on re-running the script every few days as we merge more fixes so hopefully all the "overloooked" packages will disappear "automatically" :)

**Comment by potiuk** [comment]
I Updated the list of packages - we are at 897 errrors so the numbers started to go down !

**Comment by khalidmammadov** [comment]
I'm starting to look into airflow/utils if no one started

**Comment by potiuk** [comment]
I already have a PR for that @khalidmammadov #19914

**Comment by ashb** [comment]
I'm looking at airflow/models and tests/models

**Comment by subkanthi** [comment]
Looking at airflow/providers/airbyte

**Comment by khalidmammadov** [comment]
FYI: airflow/triggers and airflow/timetables are done #20274, #20275. Although trivial fixes..

**Comment by khalidmammadov** [comment]
Anyone looking at airflow/serialization/,  airflow/sensors/, airflow/security/ and airflow/secrets? (PS.There is only one small issue in the security and secrets folder per each)

**Comment by potiuk** [comment]
> Anyone looking at airflow/serialization/, airflow/sensors/, airflow/security/ and airflow/secrets? (PS.There is only one small issue in the security and secrets folder per each)

Go ahead.

**Comment by potiuk** [comment]
BTW. After few biggish (common) fixes I plan to re-generate the list and check where we are later today :)

**Comment by subkanthi** [comment]
-  airflow/providers/exasol/hooks
-  airflow/providers/facebook/ads/hooks

https://github.com/apache/airflow/pull/20291

**Comment by subkanthi** [comment]
Started to look at 
- airflow/providers/cncf/kubernetes

**Comment by potiuk** [comment]
Updated the list - we are down to 758 errors (was 897).

**Comment by kazanzhy** [comment]
Hi team. 
I'm a little bit stuck with the Qubole operator because of the `[attr-defined]` error in Mixin.
Could you share some articles and examples of how you handle that?

**Comment by uranusjr** [comment]
Could you post the Mypy errors you can’t figure out?

**Comment by kazanzhy** [comment]
@uranusjr 
`mypy --show-error-codes --namespace-packages airflow/providers/qubole/operators/qubole_check.py`

<img width="1636" alt="Screenshot 2021-12-15 at 12 23 49" src="https://user-images.githubusercontent.com/10523218/146169180-3982f383-c840-4361-9965-0c89a5a1f2e0.png">

**Comment by potiuk** [comment]
1) the mixin has no super class. super().execute() will only work because Mixin is only applied to an Operator where BaseOperator has an execute() method. We could create a custom protocol for the mixin, but I think in this case #type: ignore with appropriate comment should be good. 

2) The get_db_hook() from BaseSQLOperator expects DBAPIHook to be returned. QBoleCheckHook does not extend DBAPIHook. Just extending from DBAPIHook should help.


Comment - this class hierarchy for Qubole is a mess. You can even see comment from @xinbinhuang which I very much agree with:

> # TODO(xinbinhuang): refactor to reduce levels of inheritance

But I think this is for another time (and likely someone from Qubole team to take care about it. Anyone from Qubole listening ;) ?

**Comment by subkanthi** [comment]
- Looking at airflow/providers/amazon/aws/hooks

**Comment by potiuk** [comment]
Today's lucky numbers: 758 -> 614. Going down fast now. List updated.

**Comment by potiuk** [comment]
I have two more waiting fore review: #20329  and #20327 (part of google) which should put the number down by a good  100  :)

**Comment by subkanthi** [comment]
Is there a way to fix this? Thanks
```
airflow/providers/cncf/kubernetes/utils/pod_launcher.py:328: error: Incompatible return value type (got "TaskInstanceState", expected "State")
                return State.QUEUED
                       ^
airflow/providers/cncf/kubernetes/utils/pod_launcher.py:331: error: Incompatible return value type (got "TaskInstanceState", expected "State")
                return State.FAILED
                       ^
airflow/providers/cncf/kubernetes/utils/pod_launcher.py:334: error: Incompatible return value type (got "TaskInstanceState", expected "State")
                return State.SUCCESS
                       ^
airflow/providers/cncf/kubernetes/utils/pod_launcher.py:336: error: Incompatible return value type (got "TaskInstanceState", expected "State")
                return State.RUNNING
                       ^
airflow/providers/cncf/kubernetes/utils/pod_launcher.py:339: error: Incompatible return value type (got "TaskInstanceState", expected "State")
                return State.FAILED

```
`    def process_status(self, job_id: str, status: str) -> State:
        """Process status information for the JOB"""
        status = status.lower()
        if status == PodStatus.PENDING:
            return State.QUEUED
        elif status == PodStatus.FAILED:
            self.log.error('Event with job id %s Failed', job_id)
            return State.FAILED
        elif status == PodStatus.SUCCEEDED:
            self.log.info('Event with job id %s Succeeded', job_id)
            return State.SUCCESS
        elif status == PodStatus.RUNNING:
            return State.RUNNING
        else:
            self.log.error('Event: Invalid state %s on job %s', status, job_id)
            return State.FAILED`

```
    # These are TaskState only
    NONE = None
    REMOVED = TaskInstanceState.REMOVED
    SCHEDULED = TaskInstanceState.SCHEDULED
    QUEUED = TaskInstanceState.QUEUED
    SHUTDOWN = TaskInstanceState.SHUTDOWN
    RESTARTING = TaskInstanceState.RESTARTING
    UP_FOR_RETRY = TaskInstanceState.UP_FOR_RETRY
    UP_FOR_RESCHEDULE = TaskInstanceState.UP_FOR_RESCHEDULE
    UPSTREAM_FAILED = TaskInstanceState.UPSTREAM_FAILED
    SKIPPED = TaskInstanceState.SKIPPED
    SENSING = TaskInstanceState.SENSING
    DEFERRED = TaskInstanceState.DEFERRED
```

**Comment by kazanzhy** [comment]
I was very surprised when saw an error in `tests/providers/snowflake`. I was sure that it's done.
Unfortunately, I fixed it in only one place and missed two ones.
Here is the PR https://github.com/apache/airflow/pull/20212/files
@khalidmammadov

**Comment by potiuk** [comment]
> Is there a way to fix this? Thanks

Good question - I used `State(State.QUEUED)` in some cases but I do not particularly like it - maybe others have other idea ?

**Comment by subkanthi** [comment]
Also looking at providers/amazon/aws/operators as the fixes in /aws/hooks affect operators

**Comment by potiuk** [comment]
I think I have all necessary changes for Google : #20358, #20329, #20234

**Comment by potiuk** [comment]
We are getting CLOSER AND CLOSER: 277 errors to go!

**Comment by khalidmammadov** [comment]
Anyone looking at airflow/providers/microsoft/?

**Comment by kazanzhy** [comment]
A think there are left hardest ones: microsoft, amazon, and google providers, and some `airflow/*` folders.
I could also take something after the revision

**Comment by potiuk** [comment]
> Anyone looking at airflow/providers/microsoft/?

Feel free!

**Comment by potiuk** [comment]
> A think there are left hardest ones: microsoft, amazon, and google providers, and some `airflow/*` folders. I could also take something after the revision

Google is almost done :). Also - we know have numbers showing how many issues are left in each package :).

**Comment by potiuk** [comment]
Taking on "apache" providers

**Comment by potiuk** [comment]
I'll take the remaining:

```
 airflow/hooks
 airflow/jobs
 airflow/macros
 airflow/operators
 airflow/example_dags
```

**Comment by subkanthi** [comment]
Question:I was on python 3.6 (when running breeze)  and I now notice some inconsistencies between 3.6 and 3.8 with regards to mypy errors, especially with typing and typing_extensions, just want to check if we should be checking multiple versions.

**Comment by potiuk** [comment]
> Question:I was on python 3.6 (when running breeze) and I now notice some inconsistencies between 3.6 and 3.8 with regards to mypy errors, especially with typing and typing_extensions, just want to check if we should be checking multiple versions.

Mypy always runs on the default version which is "3.7" at this moment (we are getting rid of Python 3.6 like - tomorrow :)

**Comment by subkanthi** [comment]
Thanks, so does that mean we don't have to worry about errors in 3.8 and just focus on 3.7.

 in the case of typing/typing_extensions, 3.6 and 3.7 are the same and 3.8 behaves differently.

```
try:
    # Literal, Protocol and TypedDict are only added to typing module starting from
    # python 3.8 we can safely remove this shim import after Airflow drops
    # support for <3.8
    from typing import Literal, Protocol, TypedDict, runtime_checkable  # type: ignore
except ImportError:
    print("Import error")
    from typing_extensions import Literal, Protocol, TypedDict, runtime_checkable  # type: ignore # noqa
```

**Comment by potiuk** [comment]
> Thanks, so does that mean we don't have to worry about errors in 3.8 and just focus on 3.7.

Yeah. 3.8 differences will be few and far between.  Not worth to check 3.7, 3.8. 3.9 seperately

**Comment by potiuk** [comment]
348 -> looks like a hydra. You cut some heads, new ones grow  :).

**Comment by potiuk** [comment]
I will take google (again) :)

**Comment by josh-fell** [comment]
#20470 should handle airflow/dag_processing but I did make some notes for review of small doubts I had.

**Comment by subkanthi** [comment]
Is there a suggested fix for missing argument,  airflow/providers/amazon/aws/example_dags have a lot of these errors.
```

airflow/providers/amazon/aws/example_dags/example_s3_bucket_tagging.py:45: error: Missing named argument "bucket_name" for
"S3CreateBucketOperator"
        create_bucket = S3CreateBucketOperator(task_id='s3_bucket_tagging_dag_create', region_name='us-east-1')

```
`with DAG(
    dag_id='s3_bucket_tagging_dag',
    schedule_interval=None,
    start_date=datetime(2021, 1, 1),
    catchup=False,
    default_args={"bucket_name": BUCKET_NAME},
    max_active_runs=1,
    tags=['example'],
) as dag:

    create_bucket = S3CreateBucketOperator(task_id='s3_bucket_tagging_dag_create', region_name='us-east-1')
`

**Comment by potiuk** [comment]
We fixed similar problems by introducing .pyi definition for operator which has the field marked as optional/having default:

Example: https://github.com/apache/airflow/pull/20422/files#diff-b203bbb3091d2f66173522a470009ba78c142564717c2168ad39d39ff7ed9a11R1 

Not ideal. But it will do for now. We will likely have to implement a custom mypy plugin to allow passing non-optional parameters by DAG's default_args and we can do that later.

**Comment by khalidmammadov** [comment]
> We fixed similar problems by introducing .pyi definition for operator which has the field marked as optional/having default:
> 
> Example: https://github.com/apache/airflow/pull/20422/files#diff-b203bbb3091d2f66173522a470009ba78c142564717c2168ad39d39ff7ed9a11R1
> 
> Not ideal. But it will do for now. We will likely have to implement a custom mypy plugin to allow passing non-optional parameters by DAG's default_args and we can do that later.

@potiuk @josh-fell I am not sure if this method fully fixes the issue. I have applied similar changed inside airflow/providers/microsoft and when I run mypy for a folder then all good but when I run it for one file only it looks like it does not acknowledge existence of companion .pyi file in the same dir. 
i.e. this is OK
`mypy --namespace-packages airflow/providers/microsoft/azure/example_dags`
but not this:
`mypy --namespace-packages airflow/providers/microsoft/azure/example_dags/example_azure_blob_to_gcs.py`

I checked `pre-commit` run approach as well it also runs on listed files and fails to detect .pyi when I replicate the command:
`pre-commit run mypy --all-files --show-diff-on-failure --color always`

Reference PR: https://github.com/apache/airflow/pull/20409

**Comment by potiuk** [comment]
> Reference PR: #20409 -- PS. I had to disable pre-commit to be able to push it (for discussion)...

Yes @khalidmammadov  - it does not work if you add example_*.pyi - we noticed that before (and it makes sense because definition of the class is elsewhere. You should add .pyi in the package where the classs is imported from. 
In #20409  you tried to change the definition in the example* and it's not going to work because stubs only work if they are next to the class they stub as I understand it

It is not perfect solution - of course, because it does not show you that the class has mandatory parameter when you use it so we will have to replace it with something better, but it will work for now.

**Comment by khalidmammadov** [comment]
> > Reference PR: #20409 -- PS. I had to disable pre-commit to be able to push it (for discussion)...
> 
> Yes @khalidmammadov - it does not work if you add `example_*.pyi` - we noticed that before (and it makes sense because definition of the class is elsewhere. You should add .pyi in the package where the classs is imported from (look at my example). In #20409 you tried to change the definition in the example* and it's not going to work because stubs only work if they are next to the class they stub as I understand it
> 
> It is not perfect solution - of course, because it does not show you that the class has mandatory parameter when you use it so we will have to replace it with something better, but it will work for now.

I see what you mean, working on it..

**Comment by subkanthi** [comment]
Question , looking at this PR, is the convention to not include the ASF source header for pyi files

https://github.com/apache/airflow/pull/20422

**Comment by potiuk** [comment]
> Question , looking at this PR, is the convention to not include the ASF source header for pyi files

This is a very good point. Technically we should do it before we reease the files in our source package. I am not sure if we want to release them, but this is another question which i think will need to be answered separately - and I think we shoud have a spearate discussion on how and whether to release them:
* we could not release them and only leave them in  the repo
* we could release them in sources or 
* we could release a separate typeshed package for those who want to use mypy https://github.com/python/typeshed

But regardless of the discussion - those fiels are manualy modified so technically speaking they should be licenced. I will update the pre-commit to include generation of headers also in those files. So even if we will not release them, it makes no harrm to add licence header.

**Comment by potiuk** [comment]
BTW. I hope we will find/implement a solution soon so that we do not have to keep most of the .pyi files - particularly those in #20422  - I see that as a tactical quick solution to a problem we have with default_args. But I hope we will get rid of them soon enough whe we will invest a bit more time in custom mypy plugin (which I think is the only "good" solution).

But there are few others - especially the "context.pyi" which I think are rather useful - not only for Airlfow developers but also airflow users who want to develop their own operators/providers/hooks and other airflow extensions. And if we have a few of those, I think we can release the .pyi files together with sources. But we need to discuss pros/cons - and licence header is a good point regardless of the outcome.

**Comment by potiuk** [comment]
cc: @ashb @uranusjr @kaxil @jedcunningham  -> I think we we need to make the decision on how/whether we release those files before 2.3. I'd say releasing all .pyi files together with airflow makes most sense.

**Comment by khalidmammadov** [comment]
Anyone looking at `tests/providers/amazon`?

**Comment by mingshi-wang** [comment]
I fixed mypy errors in airflow/decorators and then realized it was a dup effort. I'd be happy to help if there is any work not taken by anyone yet.

**Comment by potiuk** [comment]
> I fixed mypy errors in airflow/decorators and then realized it was a dup effort. I'd be happy to help if there is any work not taken by anyone yet.

See at the list in the issue - it contains a number of unresolved issues when we last checked (and check if someone has not signed up for it in the thread above). I hope to merge 3-4 more prs tomorrow and will re-run it to refresh the numbers.

**Comment by potiuk** [comment]
Hey Everyone - I have done what I've been planning to do - one big PR to update Context to be typed everywhere:
1) in Providers it's `Context` and if TYPE_CHECKING
2) elsewher it's Context and direct import

I hope it will "**just work**` with all tests :D.

**Comment by potiuk** [comment]
> I hope it will "**just work**` with all tests :D.

Looks like it is :).

**Comment by potiuk** [comment]
And "Context" merged :)

**Comment by potiuk** [comment]
Woooooohooo.. After we finally re-enable MyPy - there will be next step. New, rather interesting MyPy release:

https://mypy-lang.blogspot.com/2021/12/mypy-0930-released.html?spref=tw

Some realy cool features there that we could ue - explicit type aliases and NotRequired TypeDict elements :)

**Comment by subkanthi** [comment]
Is anyone looking at this - airflow/providers/google/cloud/operators/stackdriver.py?

**Comment by potiuk** [comment]
> Is anyone looking at this - airflow/providers/google/cloud/operators/stackdriver.py?

Feel free. but wait until I merge #20571  as there might be overlapping changes (should be in ~ 10 minutes after all tests run).

**Comment by potiuk** [comment]
Done. Refreshed the stats:

* down to 50 packages left (from 72)
* down to 325 errors left

**Comment by potiuk** [comment]
We are on definitely downward path:

* down to 39 packages
* down to 273 errrors

**Comment by potiuk** [comment]
Let me (again) take the remaining providers/google ones :)

**Comment by kaxil** [comment]
If you can, please leave amazon and probably a couple of providers for us, some of my team members wants to work on and contribute with their First PRs next Thursday

**Comment by potiuk** [comment]
> If you can, please leave amazon and probably a couple of providers for us, some of my team members wants to work on and contribute with their First PRs next Thursday

Sure  just wanted to finally(!) kill Google one :): https://github.com/apache/airflow/pull/20611

**Comment by potiuk** [comment]
* down to 38 packages
* down to 218 errrors

**Comment by subkanthi** [comment]
I briefly looked at vertica_to_mysql.py, tmpfile is accessed here in this block
```
                        cursor.execute(
                            f"LOAD DATA LOCAL INFILE '{tmpfile.name}' "
                            f"INTO TABLE {self.mysql_table} "
                            f"LINES TERMINATED BY '\r\n' ({', '.join(selected_columns)})"
                        )
```
 which is outside of the with block, not sure if this is a bug or maybe Im  missing something.

```                    with NamedTemporaryFile("w") as tmpfile:```

```
        count = 0
        with closing(vertica.get_conn()) as conn:
            with closing(conn.cursor()) as cursor:
                cursor.execute(self.sql)
                selected_columns = [d.name for d in cursor.description]

                if self.bulk_load:
                    with NamedTemporaryFile("w") as tmpfile:
                        self.log.info("Selecting rows from Vertica to local file %s...", tmpfile.name)
                        self.log.info(self.sql)

                        csv_writer = csv.writer(tmpfile, delimiter='\t', encoding='utf-8')
                        for row in cursor.iterate():
                            csv_writer.writerow(row)
                            count += 1

                        tmpfile.flush()
                else:
                    self.log.info("Selecting rows from Vertica...")
                    self.log.info(self.sql)

                    result = cursor.fetchall()
                    count = len(result)

                self.log.info("Selected rows from Vertica %s", count)

        if self.mysql_preoperator:
            self.log.info("Running MySQL preoperator...")
            mysql.run(self.mysql_preoperator)

        try:
            if self.bulk_load:
                self.log.info("Bulk inserting rows into MySQL...")
                with closing(mysql.get_conn()) as conn:
                    with closing(conn.cursor()) as cursor:
                        cursor.execute(
                            f"LOAD DATA LOCAL INFILE '{tmpfile.name}' "
                            f"INTO TABLE {self.mysql_table} "
                            f"LINES TERMINATED BY '\r\n' ({', '.join(selected_columns)})"
                        )
                        conn.commit()
                tmpfile.close()
            else:
                self.log.info("Inserting rows into MySQL...")
                mysql.insert_rows(table=self.mysql_table, rows=result, target_fields=selected_columns)
            self.log.info("Inserted rows into MySQL %s", count)
        except (MySQLdb.Error, MySQLdb.Warning):
            self.log.info("Inserted rows into MySQL 0")
            raise
```

**Comment by potiuk** [comment]
> I briefly looked at vertica_to_mysql.py, tmpfile is accessed here in this block

IT's because if it kinda distributed - opening using and closing of the tmpfille are happening only if "self.bulk_load" is True, so tmpfile will never be None there. - but MyPY is not analysing it that deeply.

I'd say we should refactor this method a bit. Let me attempt it

**Comment by khalidmammadov** [comment]
I have started looking into `airflow/providers/qubole`. Please let me know if anyone is looking at it already

**Comment by khalidmammadov** [comment]
@potiuk What's the communities view on [PEP 585](https://www.python.org/dev/peps/pep-0585/)? Looks like there will be another large refactoring needed to remove these type  aliases (Dict, List etc.) as they are deprecated (in Python 3.9) in favor of native/generic types.

**Comment by uranusjr** [comment]
The typed aliases are intentionally given a very long deprecation period and can be modified in a vastly more gradual pace. I would recommend one of the followings:

1. Keep using alises in `typing` and only start using PEP 585 definitions when Airflow drops 3.8 support (which gives the transition around two years).
2. Move the non-alised definitions when there is an official decision on PEP 649 (which allows you to safely use `from __future__ import annotation`).
3. Convert all alised annotations to stringified stdlib definitions (e.g. change `Dict[str, List[str]]` to `"dict[str, list[str]]"`; note the surrounding quotes). This works right now and does not need any future work to finish the transition, but is obviously not as pretty.

**Comment by potiuk** [comment]
I concur with @uranusjr assesment.  Even in PEP 585 it is written:

> Usefulness of this syntax before PEP 585 is limited as external tooling like Mypy does not recognize standard collections as generic. Moreover, certain features of typing like type aliases or casting require putting types outside of annotations, in runtime context. While these are relatively less common than type annotations, it's important to allow using the same type syntax in all contexts. This is why starting with Python 3.9, the following collections become generic using __class_getitem__() to parameterize contained types:

I think this will be a good exercise to make when we will prepare to drop Python 3.8 and for now using `typing` is best (ar actually when MyPy will start reporting that as a warning).

Actually (side comment), I do not really like the `abc` in Collections :). This ship has sailed already. But until the original collection.* is there this adds even more confusion (should collections.Sequence also be used as generic as well as collections.abc.Sequence ?) . The old collections.* imports will be removed in 3.9. But tt did not happen. They were finally dropped in 3.10, I believe. And only then we will have somewhat sane collections imports :). so we have indeed years until this becomes even a warning.

**Comment by khalidmammadov** [comment]
I agree with option 1 and think it will be necessary to do refactoring when Python 3.8 is retired for airflow. And I am sure there will be a lot more additions to Python during these 2 years that may affect future decisions and void any current that community can made.
Thanks for your views, I guess it's good to have in records that it was considered at this phase anyway.

**Comment by khalidmammadov** [comment]
Anyone looking at `airflow/dag_processing/processor.py`?

**Comment by potiuk** [comment]
> Anyone looking at `airflow/dag_processing/processor.py`?

Feel free!

**Comment by potiuk** [comment]
I refresehd the stats. We are:

* down to 32 packages
* down to 183 errors

@kaxil if you want your team to fix some of those you need to hurry I am afraid ;)

**Comment by kaxil** [comment]
We are doing it now :D  taking Amazon Provider for now -- so the following files:

- [ ] ( 46) airflow/providers/amazon/aws/example_dags (@phanikumv)
- [ ]  ( 14) airflow/providers/amazon/aws/hooks (@bharanidharan14)
- [ ]  ( 16) airflow/providers/amazon/aws/operators (@sunank200)
- [ ]  ( 9) airflow/providers/amazon/aws/sensors (@rajaths010494)
- [ ]  ( 1) airflow/providers/amazon/aws/transfers (Me)

**Comment by subkanthi** [comment]
> We are doing it now :D taking Amazon Provider for now -- so the following files:
> 
> * [ ]  ( 46) airflow/providers/amazon/aws/example_dags (@phanikumv)
> * [ ]   ( 14) airflow/providers/amazon/aws/hooks (@bharanidharan14)
> * [ ]   ( 16) airflow/providers/amazon/aws/operators (@sunank200)
> * [ ]   ( 9) airflow/providers/amazon/aws/sensors (@rajaths010494)
> * [ ]   ( 1) airflow/providers/amazon/aws/transfers (Me) -> [Fix MyPy issues in `airflow/providers/amazon/aws/transfers` #20708](https://github.com/apache/airflow/pull/20708)

@kaxil , heads up some work for /aws/hooks was already done here - https://github.com/apache/airflow/pull/20353 , PR is not merged yet.

**Comment by potiuk** [comment]
Todays numbers (going down fast :) ):

* 172 issues
* 29 packages

**Comment by potiuk** [comment]
* 155 issue
* 27 packages

**Comment by ashb** [comment]
I went on a bit of a bash and got all the ones from Airflow "core" - #20795

**Comment by ashb** [comment]
Hmmm, We've got a bunch of errors like this:

```airflow/providers/amazon/aws/example_dags/example_eks_templated.py:72: error:
Missing named argument "cluster_name" for "EksClusterStateSensor"
```

Which is not an error at runtime, as we use the `default_args` to provide this value (which makes sense --it's the point of default_args). I don't think there's really a way of fixing this really without adding a bunch of type ignores :(

**Comment by subkanthi** [comment]
> We fixed similar problems by introducing .pyi definition for operator which has the field marked as optional/having default:
> 
> Example: https://github.com/apache/airflow/pull/20422/files#diff-b203bbb3091d2f66173522a470009ba78c142564717c2168ad39d39ff7ed9a11R1
> 
> Not ideal. But it will do for now. We will likely have to implement a custom mypy plugin to allow passing non-optional parameters by DAG's default_args and we can do that later.

We fixed it by using pyi stub files.

From @potiuk 
We fixed similar problems by introducing .pyi definition for operator which has the field marked as optional/having default:

Example: https://github.com/apache/airflow/pull/20422/files#diff-b203bbb3091d2f66173522a470009ba78c142564717c2168ad39d39ff7ed9a11R1

Not ideal. But it will do for now. We will likely have to implement a custom mypy plugin to allow passing non-optional parameters by DAG's default_args and we can do that later.

**Comment by subkanthi** [comment]
> Hmmm, We've got a bunch of errors like this:
> 
> ```
> Missing named argument "cluster_name" for "EksClusterStateSensor"
> ```
> 
> Which is not an error at runtime, as we use the `default_args` to provide this value (which makes sense --it's the point of default_args). I don't think there's really a way of fixing this really without adding a bunch of type ignores :(


Approach we followed was using .pyi files.


We fixed similar problems by introducing .pyi definition for operator which has the field marked as optional/having default:

Example: https://github.com/apache/airflow/pull/20422/files#diff-b203bbb3091d2f66173522a470009ba78c142564717c2168ad39d39ff7ed9a11R1

Not ideal. But it will do for now. We will likely have to implement a custom mypy plugin to allow passing non-optional parameters by DAG's default_args and we can do that later.

**Comment by ashb** [comment]
> Not ideal. But it will do for now.

I'd go further than say that's not ideal and say that borderline introduces a bug that impacts our users:

Those `.pyi` files get shipped with the providers, and we're now lying about what arguments are required for the sake of which example args we happen to use in the example dags?

**Comment by ashb** [comment]
I've taken a quick look at a mypy plugin, and the `get_function_signature_hook` hook point does allow us to alter the type of the constructors. At least in theory.

However I have _no_ idea how to actually do that, nor how we could get the value of the default args out of the current DAG -- the mypy internals have next to zero documentaiton.

**Comment by potiuk** [comment]
> Those `.pyi` files get shipped with the providers, and we're now lying about what arguments are required for the sake of which example args we happen to use in the example dags?

Not really. The .pyi files are not included in released providers. They only impact 'example_dags" really - because those are the only "users" of the operators in airflow codebase. 

But what is really lying to our users are the default args in this case, 

And yea, I also looked at the mypy for that (I actually hate we have to add .pyi files, but adding #tpeignores is I think even worse).  And ye agree it's docuementation for plugins is non-existing.

I think it's 'least' of the many evils we can have there

* We could add #typeignores but they would make it part of our examples. I think very, very few of our users would use MyPY to verify their dags and adding those #typeignores in examples is generally bad idea.
* We could add #typeignores on global level (ouside of the exampleinclude) but this way we won't be really properly testing the examples (and it's useful to run MyPy on them too 

But if you come up with a better idea - I am all ears :)

**Comment by potiuk** [comment]
@ashb @uranusjr @jedcunningham (others): that actually reminded me that I we have not decided what to do with .pyi files.  As mentioned in  https://github.com/apache/airflow/issues/19891#issuecomment-1001531572

> **This is a very good point. Technically we should do it before we reease the files in our source package. I am not sure if we want to release them, but this is another question which i think will need to be answered separately - and I think we shoud have a spearate discussion on how and whether to release them:
> * we could not release them and only leave them in the repo
> * we could release them in sources or
> * we could release a separate typeshed package for those who want to use mypy https://github.com/python/typeshed

I think some of the important stub files (Context especially) we should directly include in "airflow" package. But some others (especially if we decide to leave them - for now at least - to solve some of the "default args + examples" case, we should not include). Currently we do not include any of the stub files (and we do not have a typeshed package).

We do not have a big number of stub files yet - most of them are provider's "workaround". I am quite sure we should include context.pyi - not sure about db_types/functools.

```
./airflow/utils/context.pyi
./airflow/migrations/db_types.pyi
./airflow/compat/functools.pyi
./airflow/providers/amazon/aws/operators/eks.pyi
./airflow/providers/amazon/aws/operators/s3.pyi
./airflow/providers/amazon/aws/sensors/eks.pyi
./airflow/providers/alibaba/cloud/operators/oss.pyi
./airflow/providers/microsoft/azure/operators/wasb_delete_blob.pyi
./airflow/providers/microsoft/azure/operators/cosmos.pyi
./airflow/providers/microsoft/azure/transfers/azure_blob_to_gcs.pyi
./airflow/providers/microsoft/azure/transfers/local_to_wasb.pyi
./airflow/providers/microsoft/azure/sensors/wasb.pyi
./airflow/providers/microsoft/azure/sensors/cosmos.pyi
./airflow/providers/asana/operators/asana_tasks.pyi
./airflow/providers/apache/cassandra/sensors/record.pyi
./airflow/providers/apache/cassandra/sensors/table.pyi
```

**Comment by ashb** [comment]
context.pyi - absolutely, as that one is _mostly_ for users than for Airflow itself.
db_types.pyi - not needed to be shipped, it's mostly for migrations, so developers
functools.pyi - 50/50. Not _designed_ for users, but someone could use it so I think I'd say yes to ship that one.

Providers I have no context on, but going on this discussion exclude then from the dists.

**Comment by potiuk** [comment]
> A separate typestub package is only for when the package itself can't/doesn't want to release typing, so no to that option

Agree

**Comment by uranusjr** [comment]
functools.pyi is not too useful for general users IMO; it’s there because we need to support multiple Python versions; a normal setup would generally run code against one specific Python version and can simply import the real thing, not from the compat module.

**Comment by khalidmammadov** [comment]
Is there anything left to do here?

**Comment by potiuk** [comment]
The remaining numbers
* 110 errors
* 18 packages

Updated remaining packages.

**Comment by kaxil** [comment]
I just updated the description too -- to remove ones that are solved

**Comment by kaxil** [comment]
I had just updated the list in the description , @potiuk looks like your change reverted it.. some of the ones are already solved like `imap` and the others. Check "Edit History" of the pr description

**Comment by khalidmammadov** [comment]
I am starting to look at `tests/providers/google/cloud/hooks` if no one already.

**Comment by josh-fell** [comment]
Started looking at `airflow/contrib/sensors` but let me know if anyone else has done so already.

**Comment by potiuk** [comment]
* 108 errors left

**Comment by potiuk** [comment]
Looking at azure/example_dags and amazon/example_dags.

**Comment by potiuk** [comment]
Hey everyone here in https://github.com/apache/airflow/pull/20935 I removed `eks.pyi` file - the one we introduced to "fix" default_args. It causes many more problems than it fixes - especially when we add new classes it's super confusing.

Instead I simply added this on top of the affected examples:

```
# Ignore missing args provided by default_args
# type: ignore[call-arg]
```

**Comment by potiuk** [comment]
I will remove all other "fake" .pyi files in separate PR.

**Comment by potiuk** [comment]
Removing all fakes here: https://github.com/apache/airflow/pull/20936 

Even if we loose some nice things (automated verification of "REALLY" missing arguements), this is far better than getting some errors when the .pyi files are not updated.

**Comment by josh-fell** [comment]
We have now come full circle with handling these pesky example DAGs.

**Comment by potiuk** [comment]
> We have now come full circle with handling these pesky example DAGs.

YEP

**Comment by potiuk** [comment]
Last two "biggish" https://github.com/apache/airflow/pull/21010 and https://github.com/apache/airflow/pull/20935 and we should be REALLY REALLY close.

**Comment by potiuk** [comment]
There are 12 (!) errors  left :). Can I ?

**Comment by josh-fell** [comment]
> There are 12 (!) errors left :). Can I ?

OH YEAH! Fitting you get the last one IMO since you have been driving this whole effort. Although #20930 should handle `airflow/contrib/sensors/` 🙂

**Comment by potiuk** [comment]
ALLRIGHT: 
:crossed_fingers:  everyone: https://github.com/apache/airflow/pull/21020

**Comment by potiuk** [comment]
Thanks everyone! 

MyPY is re-enabled now. Let's keep our watchful eyes over the next few days to not merge changes that are not rebased to latest main (and quickly fix those that cause main failures it it happens).

The next one when we upgrade to new MyPY release I think :)
